File: spring/actuator/src/main/java/com/hyfly/actuator/processor/ApplyJsonProcessor.java
package com.hyfly.actuator.processor;

import com.alibaba.fastjson2.JSON;
import com.hyfly.actuator.entity.message.Diagnostic;
import com.hyfly.actuator.entity.message.MessageView;
import com.hyfly.actuator.entity.message.constants.MessageLevel;
import com.hyfly.actuator.entity.message.constants.MessageType;
import lombok.Data;
import lombok.EqualsAndHashCode;
import org.apache.commons.lang3.StringUtils;

@Data
@EqualsAndHashCode(callSuper = true)
public class ApplyJsonProcessor extends BaseProcessor {
    @Override
    public void parse(String line) {
        if (completed) {
            return;
        }

        if (StringUtils.isNotBlank(line)) {
            if (line.contains("@level") && line.contains("@message") &&
                    line.contains("@module") && line.contains("@timestamp")) {
                MessageView view = JSON.parseObject(line, MessageView.class);

                if (MessageLevel.ERROR.equals(view.getLevel())) {
                    String message = view.getMessage();
                    Diagnostic diagnostic = view.getDiagnostic();
                    if (diagnostic != null) {
                        message = message + ". " + diagnostic.getDetail();
                    }

                    hasErr = true;
                    errorBuilder.append(message.trim()).append("\n");

                } else if (MessageLevel.INFO.equals(view.getLevel())) {
                    if (MessageType.CHANGE_SUMMARY.equals(view.getType())) {
                        String message = view.getMessage();
                        if (StringUtils.isNotBlank(message)) {
                            message = message.toLowerCase();
                            if (message.contains("apply complete")) {
                                completed = true;
                            }
                        }
                    }
                }
            }
        }
    }

    @Override
    public void parseError(String line) {
        hasErr = true;
        if (StringUtils.isNotBlank(line)) {
            errorBuilder.append(line.trim()).append("\n");
        }
    }
}

File: terraform/src/main/kotlin/com/hyfly/tf/processor/ApplyJsonProcessor.kt
package com.hyfly.tf.processor

import com.alibaba.fastjson2.JSON
import com.hyfly.tf.processor.entity.constants.MessageLevel
import com.hyfly.tf.processor.entity.constants.MessageType
import com.hyfly.tf.processor.entity.message.Diagnostic
import com.hyfly.tf.processor.entity.message.MessageView
import org.apache.commons.lang3.StringUtils
import java.util.*

class ApplyJsonProcessor : BaseProcessor() {

    override fun parse(line: String?) {

        if (completed) {
            return
        }

        if (StringUtils.isNotBlank(line)) {
            if (line!!.contains("@level") && line.contains("@message") &&
                line.contains("@module") && line.contains("@timestamp")
            ) {
                val view: MessageView = JSON.parseObject(line, MessageView::class.java)

                if (MessageLevel.ERROR == view.level) {
                    var message: String? = view.message
                    val diagnostic: Diagnostic? = view.diagnostic

                    if (message != null) {
                        if (diagnostic != null) {
                            message = message + ". " + diagnostic.detail
                        }

                        hasErr = true
                        errorBuilder.append(message.trim()).append("\n")
                    }
                } else if (MessageLevel.INFO == view.level) {
                    if (MessageType.CHANGE_SUMMARY == view.type) {
                        var message: String? = view.message
                        if (StringUtils.isNotBlank(message)) {
                            message = message!!.lowercase(Locale.getDefault())
                            if (message.contains("apply complete")) {
                                completed = true
                            }
                        }
                    }
                }
            }
        }
    }

    override fun parseError(line: String?) {
        hasErr = true
        if (StringUtils.isNotBlank(line)) {
            errorBuilder.append(line!!.trim()).append("\n")
        }
    }
}
